text/trace/115-add-sample-to-span-end.md

```diff
@@ -0,0 +1,182 @@
+| NOT_RECORD                | N/A (Don't Call)           | NOT_RECORD         |
+| RECORD_AND_SAMPLED        | N/A (Don't Call)           | RECORD_AND_SAMPLED |
+| RECORD                    | NOT_RECORD                 | NOT_RECORD         |
+| RECORD                    | RECORD                     | RECORD             |
+| RECORD                    | RECORD_AND_SAMPLED         | RECORD_AND_SAMPLED |
+
+## Internal details
+
+### Idempotence of samplers
+
+The calling of shouldSample multiple times can result in inconsistent results
+if the sampler is truly random. As such, the samplers will need to be idempotent
+based on the input passed. For example, a probability sampler would need to
+use data present in the passed parameters, such as the span id, to return sampling
+decisions.
+
+### Sampling
+
+## Trade-offs and mitigations
+
+See alternatives.
+
+## Prior art and alternatives
+
+Related issues:
+
+- [remove warning on trace.updateName](https://github.com/open-telemetry/opentelemetry-specification/issues/468)
+- [remove warning on trace.updateName PR](https://github.com/open-telemetry/opentelemetry-specification/pull/506)
+- [Allow samplers to be called during different moments in the Span lifetime](https://github.com/open-telemetry/opentelemetry-specification/issues/307)
+- [Sampling decision is too late to gain much performance](https://github.com/open-telemetry/opentelemetry-specification/issues/620)
+
+### Tradeoffs of multiple sampled approach
+
+- Pro: enables setting the recording decision early, which can skip additional
+       processing for instrumentations that use the isRecording field.
+- Pro: enables setting the recording decision early, which can be used
+       to set cheaper Span implementations (such as noop spans) and save on
+       processing.
+
+- Con: increases the complexity of the sampling logic slightly.
+- Con: introduces a complex matrix of final recording / sample decisions based
+       on both sample results.
+- Con: increases the processing required per span marginally (need to re-run sampling logic)
+
+#### Alternative: Always sample at the end
+
+An alternative approach is to move the existing sampling decision to the end.
+
+- Pro: reduces complexity of sampling logic
+- Pro: reduces complexity of choosing a noop span (no option to do so)
+
+- Con: every span, sampled or not, will require the same processing. This will
+       result in a net increase of system resources relative to an approach that
+       allows optimizations.
+
+### Alternative: more hooks for the sampler to update it's decision
+
+[spec #307](https://github.com/open-telemetry/opentelemetry-specification/issues/307) brings up
+a few more hooks to modify sampling decisions. This is similar to the pros and
+cons of the proposal, with an added pro of requiring more incremental effort
+on determining sampler decisions, and a con of requiring more complexity on
+the sampler's part.
```

Comment:
I described another important pro of multi-hook approach: https://github.com/open-telemetry/opentelemetry-specification/issues/307#issuecomment-642954936

Reply:
added to the otep!